File: src/main/java/cspfj/constraint/semantic/Abs.java
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractConstraint;
import cspfj.filter.RevisionHandler;
import cspfj.problem.Domain;
import cspfj.problem.Variable;

public final class Abs extends AbstractConstraint {

	private final int[] corresponding1;
	private final int[] corresponding2;
	private final int[] correspondingR;

	public Abs(final Variable result, final Variable v0) {
		super(result, v0);
		final Domain resultDom = result.getDomain();
		final Domain v0Dom = v0.getDomain();

		corresponding1 = new int[resultDom.maxSize()];
		corresponding2 = new int[resultDom.maxSize()];
		for (int i = resultDom.first(); i >= 0; i = resultDom.next(i)) {
			final int val = resultDom.value(i);
			corresponding1[i] = v0Dom.index(val);
			corresponding2[i] = v0Dom.index(-val);
		}

		correspondingR = new int[v0Dom.maxSize()];
		for (int i = v0Dom.first(); i >= 0; i = v0Dom.next(i)) {
			correspondingR[i] = resultDom.index(Math.abs(v0Dom.value(i)));
		}
	}

	@Override
	public boolean check() {
		return getValue(0) == Math.abs(getValue(1));
	}

	private static boolean valid(final int index, final Variable var) {
		return index >= 0 && var.isPresent(index);
	}

	@Override
	public boolean revise(final RevisionHandler revisator, final int reviseCount) {
		return revise0(revisator) && revise1(revisator);
	}

	private boolean revise0(final RevisionHandler revisator) {
		boolean change = false;

		final Variable var = getVariable(0);
		final Variable otherVar = getVariable(1);
		for (int i = var.getFirst(); i >= 0; i = var.getNext(i)) {
			if (!valid(corresponding1[i], otherVar)
					&& !valid(corresponding2[i], otherVar)) {
				var.remove(i);
				change = true;
			}
		}

		if (change) {
			if (var.getDomainSize() <= 0) {
				return false;
			} else if (var.getDomainSize() == 1) {
				entail();
			}
			revisator.revised(this, getVariable(0));

		}
		return true;
	}

	private boolean revise1(final RevisionHandler revisator) {
		boolean change = false;
		final Variable var = getVariable(1);
		final Variable otherVar = getVariable(0);
		for (int i = var.getFirst(); i >= 0; i = var.getNext(i)) {
			if (!valid(correspondingR[i], otherVar)) {
				var.remove(i);
				change = true;
			}
		}

		if (change) {
			if (var.getDomainSize() <= 0) {
				return false;
			} else if (var.getDomainSize() == 1) {
				entail();
			}
			revisator.revised(this, getVariable(1));
		}
		return true;
	}

	@Override
	public String toString() {
		return getVariable(0) + " = |" + getVariable(1) + "|";
	}

	@Override
	public float getEvaluation() {
		return getVariable(0).getDomainSize() * 3 / 2
				+ getVariable(1).getDomainSize();
	}
}

File: src/main/scala/cspfj/constraint/semantic/Abs.scala
package cspfj.constraint.semantic;

import cspfj.constraint.Constraint
import cspfj.Variable

final class Abs(val result: Variable, val v0: Variable) extends Constraint(Array(result, v0)) {

  //  val corresponding1 = result.dom.allValues map { v0.dom.index }
  //  val corresponding2 = result.dom.allValues map { v => v0.dom.index(-v) }
  //  val correspondingR = v0.dom.allValues map { v => result.dom.index(math.abs(v)) }

  def checkValues(t: Array[Int]) = t(0) == math.abs(t(1))

  private def valid(value: Int, variable: Variable) = {
    val index = variable.dom.index(value)
    index >= 0 && variable.dom.present(index)
  }

  def revise() = {
    var ch = result.dom.intersectVal(v0.dom.valueInterval.abs)

    if (!v0.dom.bound) {
      ch |= result.dom.filter { i =>
        val v = result.dom.value(i)
        v0.dom.presentVal(v) || v0.dom.presentVal(-v)
      }
    }

    ch |= v0.dom.intersectVal(-result.dom.lastValue, result.dom.lastValue)
    ch |= v0.dom.filter { i =>
      result.dom.presentVal(math.abs(v0.dom.value(i)))
    }

    if (ch && result.dom.size == 1 || v0.dom.size == 1) entail()

    ch
  }

  private def revise(v: Variable, f: (Int => Boolean)): Boolean = {
    val change = v.dom.filter(f)
    if (change && v.dom.size == 1) entail()
    change
  }

  override def toString = result + " = |" + v0 + "|";

  def getEvaluation = result.dom.size * 3 / 2 + v0.dom.size

  def simpleEvaluation = 1
}
